x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionRegistry.java

```diff
@@ -118,18 +118,21 @@ public boolean functionExists(String functionName) {
         // It is worth double checking if we need this copy. These are immutable anyway.
         Pattern p = Strings.hasText(pattern) ? Pattern.compile(normalize(pattern)) : null;
         return defs.entrySet().stream()
-                .filter(e -> p == null || p.matcher(e.getKey()).matches())
-                .map(e -> new FunctionDefinition(e.getKey(), emptyList(),
-                        e.getValue().clazz(), e.getValue().extractViable(), e.getValue().builder()))
-                .collect(toList());
+            .filter(e -> p == null || p.matcher(e.getKey()).matches())
+            .map(e -> cloneDefinition(e.getKey(), e.getValue()))
+            .collect(toList());
+    }
+
+    protected FunctionDefinition cloneDefinition(String name, FunctionDefinition definition) {
```

Comment:
I find this approach great, clearer intent in the immutable objects strategy. (_winking towards replaceChildren*_).